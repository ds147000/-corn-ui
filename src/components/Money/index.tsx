import React, { useMemo } from 'react'
import classNames from 'classnames'
import { Text } from '@tarojs/components'
import type { TextProps } from '@tarojs/components/types/Text'
import { formatMoney } from '../../utils'

interface MoneyProps extends TextProps {
  /** 金额类型，glod 葵花籽，coin 积分 */
  type?: 'gold' | 'coin' | 'rmb'
  /** 已经自处理过的金额 */
  amountString?: string | number
  /** 是否高亮 */
  lineThrough?: boolean
  /** 风格大小 */
  size?: 'large' | 'middle' | 'small'
  children?: string | number
}

const Money: React.FC<MoneyProps> = ({
  type = 'rmb',
  size = 'middle',
  amountString,
  lineThrough,
  children,
  className,
  ...props
}) => {
  const cln = useMemo(() => classNames('xrk-money xrk-money-' + size, className), [ size, className ])

  const value = useMemo(() => {
    if (amountString) return amountString

    if (type === 'coin') return children

    const _value = formatMoney(children || '0', undefined)
    return type === 'gold' ? `${_value}葵币` : _value
  }, [ amountString, children, type ])

  return (
    <Text {...props} className={cln}>
      {type !== 'gold' && (
        <Text
          className="xrk-money-mark"
          data-testid="money-mark"
        >
          {type === 'coin' ? 'K' : '￥'}
        </Text>
      )}
      <Text className={lineThrough ? 'line-through' : ''} data-testid="money" >{value}</Text>
    </Text>
  )
}
export default Money
