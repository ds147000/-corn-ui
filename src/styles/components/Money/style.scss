.xrk-money {
  &-mark {
    font-style: normal;
    font-size: 10px;
  }

  &-middle {
    font-size: 40px;
  }

  &-large {
    font-size: 60px;

    .xrk-money-mark {
      font-size: 30px;
    }
  }

  &-small {
    font-size: 30px;

    .xrk-money-mark {
      font-size: 20px;
    }
  }
}
