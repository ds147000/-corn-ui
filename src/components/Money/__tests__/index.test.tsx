/* eslint-disable no-magic-numbers */
/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react'
import { render } from '@testing-library/react'
import Money from '../index'


const TestFun = (
  value: any,
  type: any,
  typeExpected: string,
  valueExpected: string
): void => {
  test(`期望结果 => 类型： ${typeExpected} 金额：${valueExpected} ;`, async () => {
    const screen = render(<Money type={type}>{value}</Money>)


    if (typeExpected) {
      const Icon = screen.queryByTestId('money-mark')?.innerHTML
      expect(Icon?.trim()).toBe(typeExpected)
    }
    const valueResult = screen.getByTestId('money').innerHTML

    expect(valueResult?.trim()).toBe(valueExpected)
    expect(screen.container).toMatchSnapshot()
  })

}

describe.each([
  [ 1, '', '￥', '0.01' ],
  [ 50, '', '￥', '0.5' ],
  [ 100, '', '￥', '1' ],
  [ 4000, '', '￥', '40' ],
  [ 212321213121, '', '￥', '2123212131.21' ],
  [ 1, 'coin', 'K', '1' ],
  [ 100, 'coin', 'K', '100' ],
  [ 29104321012, 'gold', null, '291043210.12葵币' ],
  [ 1234, 'gold', null, '12.34葵币' ],
])('正常测试 <Money value={%s} type={%s} />', TestFun)


describe.each([
  [ null, '', '￥', '0' ],
  [ undefined, '', '￥', '0' ],
  [ '100', '', '￥', '1' ],
  [ '-99.2', '', '￥', '-0.99' ],
  [ -9999, '', '￥', '-99.99' ],
  [ null, 'coin', 'K', '' ],
  [ undefined, 'coin', 'K', '' ],
  [ null, 'gold', null, '0葵币' ],
  [ undefined, 'gold', null, '0葵币' ],
])('边界测试 <Money value={%s} type={%s} />', TestFun)

test('传入金额字符串 <Money value={10} amountString={100} >', () => {
  const screen = render(<Money amountString="100">10</Money>)

  const Icon =  screen.queryByTestId('money-mark')?.innerHTML
  expect(Icon?.trim()).toBe('￥')
  const valueResult = screen.getByTestId('money').innerHTML

  expect(valueResult?.trim()).toBe('100')
})

describe.each([
  [ 212321213121, '', '￥', '2123212131.21' ],
  [ 1, 'coin', 'K', '1' ],
  [ 1234, 'gold', null, '12.34葵币' ],
])('金额 <Money value={%s} type={%s} />', (
  value: any,
  type: any,
  typeExpected: string,
  valueExpected: string
) => {
  test(`金额 <Money value=${value} type=${type} lineThrough />`, () => {
    const screen = render(<Money type={type} lineThrough >{value}</Money>)

    if (typeExpected) {
      const Icon = screen.queryByTestId('money-mark')?.innerHTML
      expect(Icon?.trim()).toBe(typeExpected)
    }
    const valueResult = screen.getByTestId('money').innerHTML

    expect(valueResult?.trim()).toBe(valueExpected)
    expect(screen.container).toMatchSnapshot()
    expect(screen.container.querySelector('.line-through')).not.toBeUndefined()
  })

})
