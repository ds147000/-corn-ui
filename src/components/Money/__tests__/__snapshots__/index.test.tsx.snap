// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`正常测试 <Money value={1} type={} /> 期望结果 => 类型： ￥ 金额：0.01 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      ￥
    </span>
    <span
      class="taro-text"
      data-testid="money"
    >
      0.01
    </span>
  </span>
</div>
`;

exports[`正常测试 <Money value={1} type={coin} /> 期望结果 => 类型： K 金额：1 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      K
    </span>
    <span
      class="taro-text"
      data-testid="money"
    >
      1
    </span>
  </span>
</div>
`;

exports[`正常测试 <Money value={50} type={} /> 期望结果 => 类型： ￥ 金额：0.5 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      ￥
    </span>
    <span
      class="taro-text"
      data-testid="money"
    >
      0.5
    </span>
  </span>
</div>
`;

exports[`正常测试 <Money value={100} type={} /> 期望结果 => 类型： ￥ 金额：1 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      ￥
    </span>
    <span
      class="taro-text"
      data-testid="money"
    >
      1
    </span>
  </span>
</div>
`;

exports[`正常测试 <Money value={100} type={coin} /> 期望结果 => 类型： K 金额：100 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      K
    </span>
    <span
      class="taro-text"
      data-testid="money"
    >
      100
    </span>
  </span>
</div>
`;

exports[`正常测试 <Money value={1234} type={gold} /> 期望结果 => 类型： null 金额：12.34葵币 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text"
      data-testid="money"
    >
      12.34葵币
    </span>
  </span>
</div>
`;

exports[`正常测试 <Money value={4000} type={} /> 期望结果 => 类型： ￥ 金额：40 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      ￥
    </span>
    <span
      class="taro-text"
      data-testid="money"
    >
      40
    </span>
  </span>
</div>
`;

exports[`正常测试 <Money value={29104321012} type={gold} /> 期望结果 => 类型： null 金额：291043210.12葵币 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text"
      data-testid="money"
    >
      291043210.12葵币
    </span>
  </span>
</div>
`;

exports[`正常测试 <Money value={212321213121} type={} /> 期望结果 => 类型： ￥ 金额：2123212131.21 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      ￥
    </span>
    <span
      class="taro-text"
      data-testid="money"
    >
      2123212131.21
    </span>
  </span>
</div>
`;

exports[`边界测试 <Money value={-99.2} type={} /> 期望结果 => 类型： ￥ 金额：-0.99 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      ￥
    </span>
    <span
      class="taro-text"
      data-testid="money"
    >
      -0.99
    </span>
  </span>
</div>
`;

exports[`边界测试 <Money value={-9999} type={} /> 期望结果 => 类型： ￥ 金额：-99.99 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      ￥
    </span>
    <span
      class="taro-text"
      data-testid="money"
    >
      -99.99
    </span>
  </span>
</div>
`;

exports[`边界测试 <Money value={100} type={} /> 期望结果 => 类型： ￥ 金额：1 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      ￥
    </span>
    <span
      class="taro-text"
      data-testid="money"
    >
      1
    </span>
  </span>
</div>
`;

exports[`边界测试 <Money value={null} type={} /> 期望结果 => 类型： ￥ 金额：0 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      ￥
    </span>
    <span
      class="taro-text"
      data-testid="money"
    >
      0
    </span>
  </span>
</div>
`;

exports[`边界测试 <Money value={null} type={coin} /> 期望结果 => 类型： K 金额： ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      K
    </span>
    <span
      class="taro-text"
      data-testid="money"
    />
  </span>
</div>
`;

exports[`边界测试 <Money value={null} type={gold} /> 期望结果 => 类型： null 金额：0葵币 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text"
      data-testid="money"
    >
      0葵币
    </span>
  </span>
</div>
`;

exports[`边界测试 <Money value={undefined} type={} /> 期望结果 => 类型： ￥ 金额：0 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      ￥
    </span>
    <span
      class="taro-text"
      data-testid="money"
    >
      0
    </span>
  </span>
</div>
`;

exports[`边界测试 <Money value={undefined} type={coin} /> 期望结果 => 类型： K 金额： ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      K
    </span>
    <span
      class="taro-text"
      data-testid="money"
    />
  </span>
</div>
`;

exports[`边界测试 <Money value={undefined} type={gold} /> 期望结果 => 类型： null 金额：0葵币 ; 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text"
      data-testid="money"
    >
      0葵币
    </span>
  </span>
</div>
`;

exports[`金额 <Money value={1} type={coin} /> 金额 <Money value=1 type=coin lineThrough /> 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      K
    </span>
    <span
      class="taro-text line-through"
      data-testid="money"
    >
      1
    </span>
  </span>
</div>
`;

exports[`金额 <Money value={1234} type={gold} /> 金额 <Money value=1234 type=gold lineThrough /> 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text line-through"
      data-testid="money"
    >
      12.34葵币
    </span>
  </span>
</div>
`;

exports[`金额 <Money value={212321213121} type={} /> 金额 <Money value=212321213121 type= lineThrough /> 1`] = `
<div>
  <span
    class="taro-text xrk-money xrk-money-middle"
  >
    <span
      class="taro-text xrk-money-mark"
      data-testid="money-mark"
    >
      ￥
    </span>
    <span
      class="taro-text line-through"
      data-testid="money"
    >
      2123212131.21
    </span>
  </span>
</div>
`;
